{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OpenMS Glossary\n",
    "\n",
    "A glossary of common terms used throughout OpenMS documentation.\n",
    "\n",
    "C18  \n",
    "Octadecyl (C18) is an alkyl radical C(18)H(37) derived from an\n",
    "octadecane by removal of one hydrogen atom.\n",
    "\n",
    "CID collision-induced dissociation Collision-induced dissociation is an\n",
    "MS technique to induce fragmentation of selected ions in the gas phase,\n",
    "which are subjected to a subsequent measurement (see\n",
    "[MS2](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-MS2)).\n",
    "\n",
    "consensus features consensus feature Features from replicate experiments\n",
    "with similar retention times and m/z values are linked and considered a\n",
    "consensus feature. A consensus feature contains information on the\n",
    "common retention time and m/z values as well as intensities for each\n",
    "sample. OpenMS represents a consensus feature using the class\n",
    "[ConsensusFeature](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1ConsensusFeature.html).\n",
    "\n",
    "consensus maps consensus map A consensus map is a collection of\n",
    "[consensus features](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-consensus-features)\n",
    "identified from mass spectra across replicate experiments, usually by\n",
    "combining multiple\n",
    "[feature maps](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-feature-maps).\n",
    "One consensus map usually contains many consensus features. OpenMS\n",
    "represents a consensus map using the class\n",
    "[ConsensusMap](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1ConsensusMap.html).\n",
    "\n",
    "de novo de novo peptide sequencing A peptide's amino acid sequence is\n",
    "inferred directly from the precursor peptide mass and tandem mass\n",
    "spectrum\n",
    "([MS2](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-MS2)\n",
    "or\n",
    "[MS3](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-MS3))\n",
    "fragment ions, without comparison to a reference proteome.\n",
    "\n",
    "electrospray ionization ESI Electrospray ionization (ESI) is a technique\n",
    "used in MS to produce ions. ESI usually gives rise to peptides of charge\n",
    "2 or higher. ESI is usually coupled to Orbitrap and FTICR instruments.\n",
    "\n",
    "FASTA  \n",
    "A text-based file format for representing nucleotide or amino acid\n",
    "sequences.\n",
    "\n",
    "feature features A feature, in the OpenMS terminology, subsumes all m/z\n",
    "signals originating from a single compound at a certain charge state.\n",
    "This includes the isotope pattern and usually spans multiple spectra in\n",
    "retention time (the elution profile).\n",
    "\n",
    "feature finder FeatureFinder A FeatureFinder in pyOpenMS creates a\n",
    "[feature map](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-feature-map)\n",
    "using the MS1 spectra from a\n",
    "[peak map](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-peak-map)\n",
    "(e.g. from an mzML file) as input.\n",
    "\n",
    "feature maps feature map A feature map is a collection of\n",
    "[feature](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-feature)s\n",
    "identified from a single experiment. One feature map usually contains\n",
    "many features. OpenMS represents a feature map using the class\n",
    "[FeatureMap](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1FeatureMap.html).\n",
    "\n",
    "high performance liquid chromatography HPLC In high performance liquid\n",
    "chromatography (HPLC), analytes are dissolved in a pressurized solvent\n",
    "(mobile phase) and pumped through a solid adsorbent material (stationary\n",
    "phase) packed into a capillary column. Physicochemical properties of the\n",
    "analyte determine how strongly it interacts with the stationary phase.\n",
    "\n",
    "iTRAQ  \n",
    "Isobaric tags for relative and absolute quantitation (iTRAQ) is a MS\n",
    "based multiplexing technique designed to identify and quantify proteins\n",
    "from different samples in one single measurement.\n",
    "\n",
    "KNIME  \n",
    "An advanced workflow editor which OpenMS provides a plugin for.\n",
    "\n",
    "LC-MS LCMS\n",
    "[Liquid chromatography](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-liquid-chromatography)-coupled\n",
    "mass spectrometry.\n",
    "\n",
    "LC-MS/MS liquid chromatography coupled\n",
    "[tandem mass spectrometry](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-tandem-mass-spectrometry)\n",
    "See\n",
    "[LC-MS](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-LC-MS)\n",
    "and\n",
    "[MS2](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-MS2).\n",
    "\n",
    "liquid chromatography LC An analytical technique used to separate\n",
    "molecules of interest.\n",
    "\n",
    "LuciphorAdapter  \n",
    "Adapter for the LuciPHOr2: a site localisation tool of generic\n",
    "post-translational modifications from tandem mass spectrometry data.\n",
    "More information is available in the [OpenMS API reference\n",
    "documentation](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_LuciphorAdapter.html).\n",
    "\n",
    "MALDI  \n",
    "matrix-assisted laser desorption/ionization (MALDI) is an ionization\n",
    "technique that uses a laser energy-absorbing matrix to create ions. ESI\n",
    "usually gives rise to peptides of charge 1. MALDI is usually employed in\n",
    "combination with TOF instruments.\n",
    "\n",
    "Mass Spectrometry MS An analytical technique to measure the mass over\n",
    "charge (m/z) ratio of ions along with their abundance. This gives rise\n",
    "to a mass spectrum (with m/z on the x-axis and abundance on the y-axis).\n",
    "\n",
    "mass spectra mass spectrum A visual or numerical representation of a\n",
    "measurement from an MS instrument. A spectrum contains (usually many)\n",
    "pairs of mass-over-charge(m/z)+intensity values.\n",
    "\n",
    "MascotAdapter  \n",
    "Used to identify peptides in\n",
    "[MS2](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-MS2)\n",
    "spectra. Read more about this adapter in the [OpenMS API reference\n",
    "documentation](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_MascotAdapter.html).\n",
    "\n",
    "MSGFPlusAdapter  \n",
    "Adapter for the MS-GF+ protein identification (database search) engine.\n",
    "More information is available in the [OpenMS API reference\n",
    "documentation](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_MSGFPlusAdapter.html).\n",
    "\n",
    "MS1  \n",
    "Mass spectra of a sample where only precursor ions (i.e. no fragment\n",
    "ions) can be observed. Usually MS1 spectra are recorded to select\n",
    "targets for MS2 fragmentation.\n",
    "\n",
    "MS2 MS/MS Tandem MS is a technique where two or more mass analyzers are\n",
    "coupled together using an additional, usually destructive, reaction step\n",
    "to generate fragment ions which increases their abilities to analyse\n",
    "chemical samples.\n",
    "\n",
    "MS3  \n",
    "Multi-stage MS.\n",
    "\n",
    "mzData mzdata mzData was the first attempt by the Proteomics Standards\n",
    "Initiative (PSI) from the Human Proteome Organization (HUPO) to create a\n",
    "standardized format for MS data. This format is now deprecated, and\n",
    "replaced by mzML.\n",
    "\n",
    "mzML mzml The mzML format is an open, XML-based format for mass\n",
    "spectrometer output files, developed by the Proteomics Standard\n",
    "Initiative (PSI) with the full participation of vendors and researchers\n",
    "in order to create a single open format that would be supported by all\n",
    "software.\n",
    "\n",
    "mzXML mzxml mzXML is an open data format for storage and exchange of\n",
    "mass spectroscopy data, developed at the SPC/Institute for Systems\n",
    "Biology. This format is now deprecated, and replaced by mzML.\n",
    "\n",
    "nightly snapshot  \n",
    "Untested installers and containers which are created regularly between\n",
    "official releases and reflect the current development state.\n",
    "\n",
    "octadecyl  \n",
    "See\n",
    "[C18](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-C18).\n",
    "\n",
    "OpenMS API  \n",
    "A C++ interface that allows developers to use OpenMS core library\n",
    "classes and methods.\n",
    "\n",
    "Orbitrap orbitrap In MS, an ion trap mass analyzer consisting of an\n",
    "outer barrel-like electrode and a coaxial inner spindle-like electrode\n",
    "that traps ions in an orbital motion around the spindle. An ultra-high\n",
    "resolution MS analyzer, capable of resolving fine-isotope structure.\n",
    "\n",
    "peak maps peak map A collection of mass spectra (and/or chromatograms),\n",
    "usually sorted by retention time. Can contain spectra of one or more MS\n",
    "levels (usually level 1 and 2).\n",
    "\n",
    "peptide-spectrum match PSM PSMs A peptide-spectrum match associates a\n",
    "peptide sequence (possibly including modifications) to an MS/MS\n",
    "spectrum. This usually involves comparing the mass spectra of peptide\n",
    "fragments generated from a digested protein sample with a database of\n",
    "predicted spectra. Alternatively, this can be done using\n",
    "[de novo](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-de-novo)\n",
    "techniques (without a database, just using observed spectra).\n",
    "\n",
    "PepNovo  \n",
    "PepNovo is a de\n",
    "[de novo peptide sequencing](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-de-novo-peptide-sequencing)\n",
    "algorithm for\n",
    "[MS2](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-MS2)\n",
    "spectra.\n",
    "\n",
    "ProteoWizard  \n",
    "ProteoWizard is a set of open-source, cross-platform tools and libraries\n",
    "for proteomics data analyses. It provides a framework for unified MS\n",
    "data file access and performs standard chemistry and LCMS dataset\n",
    "computations.\n",
    "\n",
    "quadrupole  \n",
    "A low resolution MS analyzer. A mass filter allowing one mass channel at\n",
    "a time to reach the detector as the mass range is scanned.\n",
    "\n",
    "SILAC stable isotope labeling with amino acids in cell culture Stands\n",
    "for Stable isotope labeling using amino acids in cell culture.\n",
    "\n",
    "SRM  \n",
    "Selected reaction monitoring (SRM) is a MS technique for targeted small\n",
    "molecule analysis.\n",
    "\n",
    "SWATH  \n",
    "Sequential acquisition of all theoretical fragment ion spectra (SWATH)\n",
    "uses partially overlapping MS2 scans with wide isolation windows to\n",
    "capture all fragment ions in a data independent analysis (DIA).\n",
    "\n",
    "tandem mass spectrometry  \n",
    "See\n",
    "[MS2](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-MS2).\n",
    "\n",
    "time-of-flight TOF Time-of-flight (TOF) is the time taken by an object,\n",
    "particle or wave (be it acoustic, electromagnetic, etc.) to travel a\n",
    "distance through a medium. TOF analyzers can obtain good, but not\n",
    "ultra-high resolution, such as an\n",
    "[orbitrap](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-orbitrap).\n",
    "\n",
    "TMT  \n",
    "Tandem Mass Tag (TMT) is a MS based multiplexing technique designed to\n",
    "identify and quantify proteins from different samples in one single\n",
    "measurement.\n",
    "\n",
    "TOPP  \n",
    "'TOPP - The OpenMS PiPeline' is a pipeline for the analysis of HPLC-MS\n",
    "data. It consists of several small applications that can be chained to\n",
    "create analysis pipelines tailored for a specific problem. See\n",
    "[TOPP tools](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-TOPP-tools).\n",
    "\n",
    "TOPPAS  \n",
    "An assistant for GUI-driven\n",
    "[TOPP](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-TOPP)\n",
    "workflow design, build into OpenMS. See <span class=\"title-ref\">TOPPAS\n",
    "tutorial\n",
    "\\<https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPPAS_tutorial.html\\></span>\n",
    "for details.\n",
    "\n",
    "TOPP tools  \n",
    "OpenMS provides a number of applications (executable files) that are\n",
    "chainable in a pipeline/script and each process MS data. These tools are\n",
    "subdivided into different categories, such as 'File Handling' or\n",
    "'Peptide Identification'. All\n",
    "[TOPP](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-TOPP)\n",
    "tools are described in the [OpenMS API reference\n",
    "documentation](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_documentation.html).\n",
    "\n",
    "TOPPView  \n",
    "TOPPView is a viewer for MS and HPLC-MS data and shipped with every\n",
    "OpenMS release."
   ],
   "id": "11d99107-7d52-45aa-aa30-b58a29c7845a"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
