{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OpenMS Glossary\n",
    "\n",
    "A glossary of common terms used throughout OpenMS documentation.\n",
    "\n",
    "peptide-spectrum match PSM A method used in proteomics to identify\n",
    "proteins from a complex mixture. Involves comparing the mass spectra of\n",
    "peptide fragments generated from a protein sample with a database of\n",
    "predicted spectra, in order to identify the protein that produced the\n",
    "observed peptides.\n",
    "\n",
    "LC-MS LCMS\n",
    "[Liquid chromatography](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-liquid-chromatography)-coupled\n",
    "mass spectrometry.\n",
    "\n",
    "LC-MS/MS liquid chromatography coupled\n",
    "[tandem mass spectrometry](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-tandem-mass-spectrometry)\n",
    "See\n",
    "[LC-MS](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-LC-MS)\n",
    "and\n",
    "[MS2](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-MS2).\n",
    "\n",
    "liquid chromatography LC An analytical technique used to separate\n",
    "molecules of interest.\n",
    "\n",
    "FASTA  \n",
    "A text-based format for representing nucleotide or amino acid sequences.\n",
    "\n",
    "C18 octadecyl Octadecyl (C18) is an alkyl radical C(18)H(37) derived\n",
    "from an octadecane by removal of one hydrogen atom.\n",
    "\n",
    "ESI electrospray ionization Electrospray ionization (ESI) is a technique\n",
    "used in MS to produce ions.\n",
    "\n",
    "MS2 MS/MS tandem mass spectrometry Tandem MS is a technique where two or\n",
    "more mass analyzers are coupled together using an additional reaction\n",
    "step to increase their abilities to analyse chemical samples.\n",
    "\n",
    "TOF time-of-flight Time-of-flight (TOF) is the time taken by an object,\n",
    "particle of wave (be it acoustic, electromagnetic, e.t.c) to travel a\n",
    "distance through a medium.\n",
    "\n",
    "quadrupole  \n",
    "A mass filter allowing one mass channel at a time to reach the detector\n",
    "as the mass range is scanned.\n",
    "\n",
    "orbitrap  \n",
    "In MS, an ion trap mass analyzer consisting of an outer barrel-like\n",
    "electrode and a coaxial inner spindle-like electrode that traps ions in\n",
    "an orbital motion around the spindle. A high resolution MS analyzer.\n",
    "\n",
    "MS1  \n",
    "Mass spectra of a sample from a single fragmentation step.\n",
    "\n",
    "MS3  \n",
    "Multi-stage MS\n",
    "\n",
    "CID collision-induced dissociation Collision-induced dissociation is a\n",
    "MS technique to induce fragmentation of selected ions in the gas phase.\n",
    "\n",
    "TOPP  \n",
    "The OpenMS Pipeline is a set of chainable tools to create pipelines for\n",
    "mass spectrometry analysis.\n",
    "\n",
    "MSGFPlusAdapter  \n",
    "Adapter for the MS-GF+ protein identification (database search) engine.\n",
    "More information is available in the [OpenMS API reference\n",
    "documentation](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_MSGFPlusAdapter.html).\n",
    "\n",
    "LuciphorAdapter  \n",
    "Adapter for the LuciPHOr2: a site localisation tool of generic\n",
    "post-translational modifications from tandem mass spectrometry data.\n",
    "More information is available in the [OpenMS API reference\n",
    "documentation](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_LuciphorAdapter.html).\n",
    "\n",
    "TOPP tools  \n",
    "OpenMS provides a number of functions that process MS data called\n",
    "[TOPP](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-TOPP)\n",
    "tools. All\n",
    "[TOPP](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-TOPP)\n",
    "tools are described in the [OpenMS API reference\n",
    "documentation](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_documentation.html).\n",
    "\n",
    "UTILS  \n",
    "Besides\n",
    "[TOPP tools](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-TOPP-tools),\n",
    "OpenMS offers a range of other tools. They are not included in\n",
    "[TOPP](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-TOPP)\n",
    "as they are not part of typical analysis pipelines. More information is\n",
    "present in [OpenMS API reference\n",
    "documentation](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/UTILS_documentation.html).\n",
    "\n",
    "TOPPView  \n",
    "TOPPView is a viewer for MS and HPLC-MS data.\n",
    "\n",
    "nightly snapshot  \n",
    "Untested installers and containers are known as the nightly snapshot.\n",
    "\n",
    "MascotAdapter  \n",
    "Used to identifies peptides in MS2 spectra. Read more about\n",
    "MascotAdapter in the [OpenMS API reference\n",
    "documentation](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/TOPP_MascotAdapter.html).\n",
    "\n",
    "high performance liquid chromatography HPLC In high performance liquid\n",
    "chromatography (HPLC), analytes are dissolved in a pressurized solvent\n",
    "(mobile phase) and pumped through a solid adsorbent material (stationary\n",
    "phase) packed into a capillary column. Physicochemical properties of the\n",
    "analyte determine how strongly it interacts with the stationary phase.\n",
    "\n",
    "mzML mzml The mzML format is an open, XML-based format for mass\n",
    "spectrometer output files, developed with the full participation of\n",
    "vendors and researchers in order to create a single open format that\n",
    "would be supported by all software.\n",
    "\n",
    "mzData mzdata mzData was the first attempt by the Proteomics Standards\n",
    "Initiative (PSI) from the Human Proteome Organization (HUPO) to create a\n",
    "standardized format for MS data. This format is now deprecated, and\n",
    "replaced by mzML.\n",
    "\n",
    "mzXML mzxml mzXML is an open data format for storage and exchange of\n",
    "mass spectroscopy data, developed at the SPC/Institute for Systems\n",
    "Biology.\n",
    "\n",
    "ProteoWizard  \n",
    "ProteoWizard is a set of open-source, cross-platform tools and libraries\n",
    "for proteomics data analyses. It provides a framework for unified MS\n",
    "data file access and performs standard chemistry and LCMS dataset\n",
    "computations.\n",
    "\n",
    "PepNovo  \n",
    "PepNovo is a de\n",
    "[de novo peptide sequencing](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-de-novo-peptide-sequencing)\n",
    "algorithm for\n",
    "[MS2](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-MS2)\n",
    "spectra.\n",
    "\n",
    "de novo peptide sequencing  \n",
    "A peptideâ€™s amino acid sequence is inferred directly from the precursor\n",
    "peptide mass and tandem mass spectrum\n",
    "([MS2](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-MS2)\n",
    "or\n",
    "[MS3](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-MS3))\n",
    "fragment ions, without comparison to a reference proteome.\n",
    "\n",
    "TOPPAS  \n",
    "An assistant for GUI-driven\n",
    "[TOPP](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-TOPP)\n",
    "workflow design. It is recommended to use OpenMS through the KNIME\n",
    "plugins.\n",
    "\n",
    "KNIME  \n",
    "An advanced workflow editor which OpenMS provides a plugin for.\n",
    "\n",
    "SILAC stable isotope labeling with amino acids in cell culture Stands\n",
    "for Stable isotope labeling using amino acids in cell culture.\n",
    "\n",
    "iTRAQ  \n",
    "Stands for isobaric tags for relative and absolute quantitation.\n",
    "\n",
    "TMT  \n",
    "Tandem Mass Tag (TMT) is a MS based system designed to identify and\n",
    "quantify proteins in different samples.\n",
    "\n",
    "SRM  \n",
    "Selected reaction monitoring is a MS technique for small molecule\n",
    "analysis.\n",
    "\n",
    "SWATH  \n",
    "Stands for sequential acquisition of all theoretical fragment ion\n",
    "spectra.\n",
    "\n",
    "OpenMS API  \n",
    "An interface that allows developers to use OpenMS core library classes\n",
    "and methods.\n",
    "\n",
    "feature maps feature map A feature map is a collection of features\n",
    "identified in a mass spectrum from a single experiment. One feature map\n",
    "can contain many features. OpenMS represents a feature map using the\n",
    "class\n",
    "[FeatureMap](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1FeatureMap.html).\n",
    "\n",
    "consensus features consensus feature Features from replicate experiments\n",
    "with similar retention times and m/z values are linked and considered a\n",
    "consensus feature. A consensus feature contains information on the\n",
    "common retention time and m/z values as well as intensities for each\n",
    "sample. OpenMS represents a consensus feature using the class\n",
    "[ConsensusFeature](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1ConsensusFeature.html).\n",
    "\n",
    "consensus maps consensus map A consensus map is a collection of\n",
    "[consensus features](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-consensus-features)\n",
    "identified from mass spectra across replicate experiments. One consensus\n",
    "map can contain many consensus features. OpenMS represents a consensus\n",
    "map using the class\n",
    "[ConsensusMap](https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/nightly/html/classOpenMS_1_1ConsensusMap.html)."
   ],
   "id": "16a386d7-039d-4d5b-9132-cfcbfeb25796"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
