{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary\n",
    "\n",
    "pyOpenMS is an open-source Python library for\n",
    "[mass spectrometry](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-Mass-spectrometry),\n",
    "specifically for the analysis of proteomics and metabolomics data in\n",
    "Python. pyOpenMS implements a set of Python bindings to the OpenMS\n",
    "library for computational\n",
    "[mass spectrometry](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-Mass-spectrometry)\n",
    "and is available for Windows, Linux and macOS.\n",
    "\n",
    "pyOpenMS provides functionality that is commonly used in computational\n",
    "mass spectrometry. The pyOpenMS package contains Python bindings for a\n",
    "large part of the [OpenMS](openms.de) library for\n",
    "[mass spectrometry](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-Mass-spectrometry)\n",
    "based proteomics. It thus provides easy access to a feature-rich,\n",
    "open-source algorithm library for mass-spectrometry based proteomics\n",
    "analysis.\n",
    "\n",
    "pyOpenMS facilitates the execution of common tasks in proteomics (and\n",
    "other fields of mass spectrometry) such as\n",
    "\n",
    "-   File handling\n",
    "    ([mzXML](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-mzXML),\n",
    "    [mzML](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-mzML),\n",
    "    TraML, mzTab,\n",
    "    [FASTA](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-FASTA),\n",
    "    pepxml, protxml, mzIdentML among others)\n",
    "-   Chemistry (mass calculation, peptide fragmentation, isotopic\n",
    "    abundances)\n",
    "-   Signal processing (smoothing, filtering, de-isotoping, retention\n",
    "    time correction and peak-picking)\n",
    "-   Identification analysis (including peptide search, PTM analysis,\n",
    "    cross-linked analytes, FDR control, RNA oligonucleotide search and\n",
    "    small molecule search tools)\n",
    "-   Quantitative analysis (including label-free, metabolomics,\n",
    "    [SILAC](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-SILAC),\n",
    "    [iTRAQ](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-iTRAQ)\n",
    "    and\n",
    "    [SWATH](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-SWATH)/DIA\n",
    "    analysis tools)\n",
    "-   Chromatogram analysis (chromatographic peak picking, smoothing,\n",
    "    elution profiles and peak scoring for\n",
    "    [SRM](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-SRM)/MRM/PRM/[SWATH](https://pyopenms.readthedocs.io/en/latest/user_guide/glossary.html#term-SWATH)/DIA\n",
    "    data)\n",
    "-   Interaction with common tools in proteomics and metabolomics:\n",
    "    -   Search engines such as Comet, Mascot, MSGF+, MSFragger,\n",
    "        SpectraST, XTandem\n",
    "    -   Post-processing tools such as Percolator, MSstats, Fido\n",
    "    -   Metabolomics tools such as SIRIUS, CSI:FingerId\n",
    "\n",
    "# User Guide\n",
    "\n",
    "Information about installing and using pyopenms.\n",
    "\n",
    "user_guide/index\n",
    "\n",
    "# API documentation\n",
    "\n",
    "Documentation on modules, functions, classes of this package.\n",
    "\n",
    "apidocs/index\n",
    "\n",
    "# Community and contribution guide\n",
    "\n",
    "Information about the community behind this package and how you can\n",
    "contribute.\n",
    "\n",
    "community/index\n",
    "\n",
    "C++ library \\<<https://openms.de/documentation/html/index.html>\\>"
   ],
   "id": "9d056b3e-70e4-4749-837a-c3f133a6aebd"
  }
 ],
 "nbformat": 4,
 "nbformat_minor": 5,
 "metadata": {}
}
