{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "bf63ecd0",
   "metadata": {},
   "source": [
    "$$\n",
    "\\def\\CC{\\bf C}\n",
    "\\def\\QQ{\\bf Q}\n",
    "\\def\\RR{\\bf R}\n",
    "\\def\\ZZ{\\bf Z}\n",
    "\\def\\NN{\\bf N}\n",
    "$$\n",
    "# Summary\n",
    "\n",
    "pyOpenMS is an open-source Python library for mass spectrometry,\n",
    "specifically for the analysis of proteomics and metabolomics data in\n",
    "Python. pyOpenMS implements a set of Python bindings to the OpenMS\n",
    "library for computational mass spectrometry and is available for\n",
    "Windows, Linux and OSX.\n",
    "\n",
    "PyOpenMS provides functionality that is commonly used in computational\n",
    "mass spectrometry. The pyOpenMS package contains Python bindings for a\n",
    "large part of the OpenMS library (<http://www.open-ms.de>) for mass\n",
    "spectrometry based proteomics. It thus provides facile access to a\n",
    "feature-rich, open-source algorithm library for mass-spectrometry based\n",
    "proteomics analysis.\n",
    "\n",
    "pyOpenMS facilitates the execution of common tasks in protoemics (and\n",
    "other mass spectrometric fields) such as\n",
    "\n",
    "-   file handling (mzXML, mzML, TraML, mzTab, fasta, pepxml, protxml,\n",
    "    mzIdentML among others)\n",
    "-   chemistry (mass calculation, peptide fragmentation, isotopic\n",
    "    abundances)\n",
    "-   signal processing (smoothing, filtering, de-isotoping, retention\n",
    "    time correction and peak-picking)\n",
    "-   identification analysis (including peptide search, PTM analysis,\n",
    "    Cross-linked analytes, FDR control, RNA oligonucleotide search and\n",
    "    small molecule search tools)\n",
    "-   quantitative analysis (including label-free, metabolomics, SILAC,\n",
    "    iTRAQ and SWATH/DIA analysis tools)\n",
    "-   chromatogram analysis (chromatographic peak picking, smoothing,\n",
    "    elution profiles and peak scoring for SRM/MRM/PRM/SWATH/DIA data)\n",
    "-   interaction with common tools in proteomics and metabolomics\n",
    "    -   search engines such as Comet, Crux, Mascot, MSGFPlus, MSFragger,\n",
    "        Myrimatch, OMSSA, Sequest, SpectraST, XTandem\n",
    "    -   post-processing tools such as percolator, MSStats, Fido\n",
    "    -   metabolomics tools such as SIRIUS, CSI:FingerId\n",
    "\n",
    "Please see the appendix of the official [pyOpenMS\n",
    "Manual](http://proteomics.ethz.ch/pyOpenMS_Manual.pdf) for a complete\n",
    "documentation of the pyOpenMS API and all wrapped classes.\n",
    "\n",
    "Note: the current documentation relates to the 2.6.0 release of\n",
    "pyOpenMS.\n",
    "\n",
    "installation\n",
    "\n",
    "getting\\_started file\\_handling other\\_file\\_handling\n",
    "\n",
    "datastructures chemistry aasequences nasequences\n",
    "theoreticalspectrumgenerator digestion datastructures\\_id\n",
    "datastructures\\_quant\n",
    "\n",
    "data\\_manipulation\n",
    "\n",
    "parameter\\_handling algorithms smoothing mass\\_decomposition deisotoping\n",
    "feature\\_detection peptide\\_search chromatographic\\_analysis\n",
    "metabolomics\\_targeted\\_feature\\_extraction\n",
    "\n",
    "mzMLFileFormat\n",
    "\n",
    "pyopenms\\_in\\_r\n",
    "\n",
    "build\\_from\\_source wrap\\_classes\n",
    "\n",
    "# Indices and tables\n",
    "\n",
    "-   [genindex](genindex.ipynb)\n",
    "\n",
    "[![Launch Binder](./img/launch_binder.jpg)](https://mybinder.org/v2/gh/OpenMS/pyopenms-extra/master+ipynb?urlpath=lab/tree/docs/source/index.ipynb)"
   ]
  }
 ],
 "metadata": {},
 "nbformat": 4,
 "nbformat_minor": 5
}
