

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Summary &mdash; pyOpenMS 2.5.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="pyOpenMS Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="#" class="icon icon-home"> pyOpenMS
          

          
          </a>

          
            
            
              <div class="version">
                2.5.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">pyOpenMS Installation</a></li>
</ul>
<p class="caption"><span class="caption-text">First steps</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="file_handling.html">Reading Raw MS data</a></li>
<li class="toctree-l1"><a class="reference internal" href="mzMLFileFormat.html">mzML files</a></li>
<li class="toctree-l1"><a class="reference internal" href="other_file_handling.html">Other MS data formats</a></li>
</ul>
<p class="caption"><span class="caption-text">Mass Spectrometry Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="datastructures.html">MS Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="chemistry.html">Chemistry</a></li>
<li class="toctree-l1"><a class="reference internal" href="aasequences.html">Peptides and Proteins</a></li>
<li class="toctree-l1"><a class="reference internal" href="nasequences.html">Oligonucleotides: RNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="theoreticalspectrumgenerator.html">TheoreticalSpectrumGenerator</a></li>
<li class="toctree-l1"><a class="reference internal" href="digestion.html">Digestion</a></li>
<li class="toctree-l1"><a class="reference internal" href="datastructures_id.html">Identification Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="datastructures_quant.html">Quantitative Data</a></li>
</ul>
<p class="caption"><span class="caption-text">Simple Data Manipulation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="data_manipulation.html">Simple Data Manipulation</a></li>
</ul>
<p class="caption"><span class="caption-text">OpenMS Algorithms</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="parameter_handling.html">Parameter Handling</a></li>
<li class="toctree-l1"><a class="reference internal" href="algorithms.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="smoothing.html">Smoothing</a></li>
<li class="toctree-l1"><a class="reference internal" href="mass_decomposition.html">Mass Decomposition</a></li>
<li class="toctree-l1"><a class="reference internal" href="deisotoping.html">Charge and Isotope Deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="feature_detection.html">Feature Detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="peptide_search.html">Peptide Search</a></li>
<li class="toctree-l1"><a class="reference internal" href="chromatographic_analysis.html">Chromagraphic Analysis</a></li>
</ul>
<p class="caption"><span class="caption-text">R language</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pyopenms_in_r.html">pyOpenMS in R</a></li>
</ul>
<p class="caption"><span class="caption-text">Developers</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="build_from_source.html">Build from source</a></li>
<li class="toctree-l1"><a class="reference internal" href="wrap_classes.html">Wrapping Workflow and wrapping new Classes</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">pyOpenMS</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#">Docs</a> &raquo;</li>
        
      <li>Summary</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="summary">
<h1>Summary<a class="headerlink" href="#summary" title="Permalink to this headline">¶</a></h1>
<p>pyOpenMS is an open-source Python library for mass spectrometry, specifically for the analysis of proteomics and metabolomics data in Python. pyOpenMS implements a set of Python bindings to the OpenMS library for computational mass spectrometry and is available for Windows, Linux and OSX.</p>
<p>PyOpenMS provides functionality that is commonly used in computational mass
spectrometry.  The pyOpenMS package contains Python bindings for a large part of the OpenMS
library (<a class="reference external" href="http://www.open-ms.de">http://www.open-ms.de</a>) for mass spectrometry based proteomics. It thus
provides facile access to a feature-rich, open-source algorithm
library for mass-spectrometry based proteomics analysis.</p>
<p>pyOpenMS facilitates the execution of common tasks in protoemics (and other mass spectrometric fields) such as</p>
<ul class="simple">
<li>file handling (mzXML, mzML, TraML, mzTab, fasta, pepxml, protxml, mzIdentML among others)</li>
<li>chemistry (mass calculation, peptide fragmentation, isotopic abundances)</li>
<li>signal processing (smoothing, filtering, de-isotoping, retention time correction and peak-picking)</li>
<li>identification analysis (including peptide search, PTM analysis, Cross-linked analytes, FDR control, RNA oligonucleotide search and small molecule search tools)</li>
<li>quantitative analysis (including label-free, metabolomics, SILAC, iTRAQ and SWATH/DIA analysis tools)</li>
<li>chromatogram analysis (chromatographic peak picking, smoothing, elution profiles and peak scoring for SRM/MRM/PRM/SWATH/DIA data)</li>
<li>interaction with common tools in proteomics and metabolomics<ul>
<li>search engines such as Comet, Crux, Mascot, MSGFPlus, MSFragger, Myrimatch, OMSSA, Sequest, SpectraST, XTandem</li>
<li>post-processing tools such as percolator, MSStats, Fido</li>
<li>metabolomics tools such as SIRIUS, CSI:FingerId</li>
</ul>
</li>
</ul>
<p>Please see the appendix of the official <a class="reference external" href="http://proteomics.ethz.ch/pyOpenMS_Manual.pdf">pyOpenMS Manual</a> for a complete documentation
of the pyOpenMS API and all wrapped classes.</p>
<p>Note: the current documentation relates to the 2.5.0 release of pyOpenMS.</p>
<div class="toctree-wrapper compound">
<p class="caption"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">pyOpenMS Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="installation.html#binaries">Binaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#source">Source</a></li>
<li class="toctree-l2"><a class="reference internal" href="installation.html#wrap-classes">Wrap Classes</a></li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<p class="caption"><span class="caption-text">First steps</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a><ul>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#import-pyopenms">Import pyopenms</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#getting-help">Getting help</a></li>
<li class="toctree-l2"><a class="reference internal" href="getting_started.html#first-look-at-data">First look at data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="file_handling.html">Reading Raw MS data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="file_handling.html#mzml-files-in-memory">mzML files in memory</a></li>
<li class="toctree-l2"><a class="reference internal" href="file_handling.html#indexed-mzml-files">indexed mzML files</a></li>
<li class="toctree-l2"><a class="reference internal" href="file_handling.html#mzml-files-as-streams">mzML files as streams</a></li>
<li class="toctree-l2"><a class="reference internal" href="file_handling.html#cached-mzml-files">cached mzML files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="mzMLFileFormat.html">mzML files</a><ul>
<li class="toctree-l2"><a class="reference internal" href="mzMLFileFormat.html#binary-encoding">Binary encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="mzMLFileFormat.html#id1">Base64 encoding</a></li>
<li class="toctree-l2"><a class="reference internal" href="mzMLFileFormat.html#numpress-encoding">numpress encoding</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="other_file_handling.html">Other MS data formats</a><ul>
<li class="toctree-l2"><a class="reference internal" href="other_file_handling.html#identification-data-idxml-mzidentml-pepxml-protxml">Identification data (idXML, mzIdentML, pepXML, protXML)</a></li>
<li class="toctree-l2"><a class="reference internal" href="other_file_handling.html#quantiative-data-featurexml-consensusxml">Quantiative data (featureXML, consensusXML)</a></li>
<li class="toctree-l2"><a class="reference internal" href="other_file_handling.html#transition-data-traml">Transition data (TraML)</a></li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<p class="caption"><span class="caption-text">Mass Spectrometry Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="datastructures.html">MS Data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="datastructures.html#spectrum">Spectrum</a></li>
<li class="toctree-l2"><a class="reference internal" href="datastructures.html#lc-ms-ms-experiment">LC-MS/MS Experiment</a></li>
<li class="toctree-l2"><a class="reference internal" href="datastructures.html#chromatogram">Chromatogram</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chemistry.html">Chemistry</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chemistry.html#constants">Constants</a></li>
<li class="toctree-l2"><a class="reference internal" href="chemistry.html#elements">Elements</a></li>
<li class="toctree-l2"><a class="reference internal" href="chemistry.html#molecular-formulae">Molecular Formulae</a></li>
<li class="toctree-l2"><a class="reference internal" href="chemistry.html#isotopic-distributions">Isotopic Distributions</a></li>
<li class="toctree-l2"><a class="reference internal" href="chemistry.html#amino-acids">Amino Acids</a></li>
<li class="toctree-l2"><a class="reference internal" href="chemistry.html#amino-acid-modifications">Amino Acid Modifications</a></li>
<li class="toctree-l2"><a class="reference internal" href="chemistry.html#ribonucleotides">Ribonucleotides</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="aasequences.html">Peptides and Proteins</a><ul>
<li class="toctree-l2"><a class="reference internal" href="aasequences.html#amino-acid-sequences">Amino Acid Sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="aasequences.html#modified-sequences">Modified Sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="aasequences.html#proteins">Proteins</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="nasequences.html">Oligonucleotides: RNA</a><ul>
<li class="toctree-l2"><a class="reference internal" href="nasequences.html#nucleic-acid-sequences">Nucleic Acid Sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="nasequences.html#modified-oligonucleotides">Modified oligonucleotides</a></li>
<li class="toctree-l2"><a class="reference internal" href="nasequences.html#dna-rna-and-protein">DNA, RNA and Protein</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="theoreticalspectrumgenerator.html">TheoreticalSpectrumGenerator</a><ul>
<li class="toctree-l2"><a class="reference internal" href="theoreticalspectrumgenerator.html#y-ion-spectrum">Y-ion spectrum</a></li>
<li class="toctree-l2"><a class="reference internal" href="theoreticalspectrumgenerator.html#full-fragment-ion-spectrum">Full fragment ion spectrum</a></li>
<li class="toctree-l2"><a class="reference internal" href="theoreticalspectrumgenerator.html#visualization">Visualization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="digestion.html">Digestion</a><ul>
<li class="toctree-l2"><a class="reference internal" href="digestion.html#proteolytic-digestion-with-trypsin">Proteolytic Digestion with Trypsin</a></li>
<li class="toctree-l2"><a class="reference internal" href="digestion.html#proteolytic-digestion-with-lys-c">Proteolytic Digestion with Lys-C</a></li>
<li class="toctree-l2"><a class="reference internal" href="digestion.html#oligonucleotide-digestion">Oligonucleotide Digestion</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="datastructures_id.html">Identification Data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="datastructures_id.html#proteinidentification">ProteinIdentification</a></li>
<li class="toctree-l2"><a class="reference internal" href="datastructures_id.html#peptideidentification">PeptideIdentification</a></li>
<li class="toctree-l2"><a class="reference internal" href="datastructures_id.html#storage-on-disk">Storage on disk</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="datastructures_quant.html">Quantitative Data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="datastructures_quant.html#feature">Feature</a></li>
<li class="toctree-l2"><a class="reference internal" href="datastructures_quant.html#featuremap">FeatureMap</a></li>
<li class="toctree-l2"><a class="reference internal" href="datastructures_quant.html#consensusfeature">ConsensusFeature</a></li>
<li class="toctree-l2"><a class="reference internal" href="datastructures_quant.html#consensusmap">ConsensusMap</a></li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<p class="caption"><span class="caption-text">Simple Data Manipulation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="data_manipulation.html">Simple Data Manipulation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="data_manipulation.html#filtering-spectra">Filtering Spectra</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_manipulation.html#filtering-spectra-and-peaks">Filtering Spectra and Peaks</a></li>
<li class="toctree-l2"><a class="reference internal" href="data_manipulation.html#memory-management">Memory management</a></li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<p class="caption"><span class="caption-text">OpenMS Algorithms</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="parameter_handling.html">Parameter Handling</a></li>
<li class="toctree-l1"><a class="reference internal" href="algorithms.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="smoothing.html">Smoothing</a></li>
<li class="toctree-l1"><a class="reference internal" href="mass_decomposition.html">Mass Decomposition</a><ul>
<li class="toctree-l2"><a class="reference internal" href="mass_decomposition.html#fragment-mass-to-amino-acid-composition">Fragment mass to amino acid composition</a></li>
<li class="toctree-l2"><a class="reference internal" href="mass_decomposition.html#naive-algorithm">Naive algorithm</a></li>
<li class="toctree-l2"><a class="reference internal" href="mass_decomposition.html#stand-alone-program">Stand-alone Program</a></li>
<li class="toctree-l2"><a class="reference internal" href="mass_decomposition.html#spectrum-tagger">Spectrum Tagger</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="deisotoping.html">Charge and Isotope Deconvolution</a><ul>
<li class="toctree-l2"><a class="reference internal" href="deisotoping.html#single-peak-example">Single peak example</a></li>
<li class="toctree-l2"><a class="reference internal" href="deisotoping.html#full-spectral-de-isotoping">Full spectral de-isotoping</a></li>
<li class="toctree-l2"><a class="reference internal" href="deisotoping.html#visualization">Visualization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="feature_detection.html">Feature Detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="peptide_search.html">Peptide Search</a><ul>
<li class="toctree-l2"><a class="reference internal" href="peptide_search.html#simplesearch">SimpleSearch</a></li>
<li class="toctree-l2"><a class="reference internal" href="peptide_search.html#psm-inspection">PSM inspection</a></li>
<li class="toctree-l2"><a class="reference internal" href="peptide_search.html#visualization">Visualization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chromatographic_analysis.html">Chromagraphic Analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="chromatographic_analysis.html#peak-detection">Peak Detection</a></li>
<li class="toctree-l2"><a class="reference internal" href="chromatographic_analysis.html#visualization">Visualization</a></li>
<li class="toctree-l2"><a class="reference internal" href="chromatographic_analysis.html#smoothing">Smoothing</a></li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<p class="caption"><span class="caption-text">R language</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pyopenms_in_r.html">pyOpenMS in R</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pyopenms_in_r.html#install-the-reticulate-r-package">Install the “reticulate” R package</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyopenms_in_r.html#import-pyopenms-in-r">Import pyopenms in R</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyopenms_in_r.html#getting-help">Getting help</a></li>
<li class="toctree-l2"><a class="reference internal" href="pyopenms_in_r.html#an-example-use-case">An example use case</a></li>
</ul>
</li>
</ul>
</div>
<div class="toctree-wrapper compound">
<p class="caption"><span class="caption-text">Developers</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="build_from_source.html">Build from source</a><ul>
<li class="toctree-l2"><a class="reference internal" href="build_from_source.html#further-questions">Further questions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="wrap_classes.html">Wrapping Workflow and wrapping new Classes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="wrap_classes.html#how-pyopenms-wraps-python-classes">How pyOpenMS wraps Python classes</a></li>
<li class="toctree-l2"><a class="reference internal" href="wrap_classes.html#how-to-wrap-new-methods-in-existing-classes">How to wrap new methods in existing classes</a></li>
<li class="toctree-l2"><a class="reference internal" href="wrap_classes.html#how-to-wrap-new-classes">How to wrap new classes</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="indices-and-tables">
<h1>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></li>
</ul>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="installation.html" class="btn btn-neutral float-right" title="pyOpenMS Installation" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, OpenMS Team

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>