

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Peptide Search &mdash; pyOpenMS 2.5.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Chromagraphic Analysis" href="chromatographic_analysis.html" />
    <link rel="prev" title="Feature Detection" href="feature_detection.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> pyOpenMS
          

          
          </a>

          
            
            
              <div class="version">
                2.5.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">pyOpenMS Installation</a></li>
</ul>
<p class="caption"><span class="caption-text">First steps</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="file_handling.html">Reading Raw MS data</a></li>
<li class="toctree-l1"><a class="reference internal" href="mzMLFileFormat.html">mzML files</a></li>
<li class="toctree-l1"><a class="reference internal" href="other_file_handling.html">Other MS data formats</a></li>
</ul>
<p class="caption"><span class="caption-text">Mass Spectrometry Concepts</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="datastructures.html">MS Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="chemistry.html">Chemistry</a></li>
<li class="toctree-l1"><a class="reference internal" href="aasequences.html">Peptides and Proteins</a></li>
<li class="toctree-l1"><a class="reference internal" href="nasequences.html">Oligonucleotides: RNA</a></li>
<li class="toctree-l1"><a class="reference internal" href="theoreticalspectrumgenerator.html">TheoreticalSpectrumGenerator</a></li>
<li class="toctree-l1"><a class="reference internal" href="digestion.html">Digestion</a></li>
<li class="toctree-l1"><a class="reference internal" href="datastructures_id.html">Identification Data</a></li>
<li class="toctree-l1"><a class="reference internal" href="datastructures_quant.html">Quantitative Data</a></li>
</ul>
<p class="caption"><span class="caption-text">Simple Data Manipulation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="data_manipulation.html">Simple Data Manipulation</a></li>
</ul>
<p class="caption"><span class="caption-text">OpenMS Algorithms</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="parameter_handling.html">Parameter Handling</a></li>
<li class="toctree-l1"><a class="reference internal" href="algorithms.html">Algorithms</a></li>
<li class="toctree-l1"><a class="reference internal" href="smoothing.html">Smoothing</a></li>
<li class="toctree-l1"><a class="reference internal" href="mass_decomposition.html">Mass Decomposition</a></li>
<li class="toctree-l1"><a class="reference internal" href="deisotoping.html">Charge and Isotope Deconvolution</a></li>
<li class="toctree-l1"><a class="reference internal" href="feature_detection.html">Feature Detection</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Peptide Search</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#simplesearch">SimpleSearch</a></li>
<li class="toctree-l2"><a class="reference internal" href="#psm-inspection">PSM inspection</a></li>
<li class="toctree-l2"><a class="reference internal" href="#visualization">Visualization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="chromatographic_analysis.html">Chromagraphic Analysis</a></li>
</ul>
<p class="caption"><span class="caption-text">R language</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pyopenms_in_r.html">pyOpenMS in R</a></li>
</ul>
<p class="caption"><span class="caption-text">Developers</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="build_from_source.html">Build from source</a></li>
<li class="toctree-l1"><a class="reference internal" href="wrap_classes.html">Wrapping Workflow and wrapping new Classes</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pyOpenMS</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Peptide Search</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/peptide_search.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="peptide-search">
<h1>Peptide Search<a class="headerlink" href="#peptide-search" title="Permalink to this headline">¶</a></h1>
<p>In MS-based proteomics, fragment ion spectra (MS2 spectra) are often
interpreted by comparing them against a theoretical set of spectra generated
from a FASTA database. OpenMS contains a (simple) implementation of such a
“search engine” that compares experimental spectra against theoretical spectra
generated from an enzymatic or chemical digest of a proteome (e.g. tryptic
digest).</p>
<div class="section" id="simplesearch">
<h2>SimpleSearch<a class="headerlink" href="#simplesearch" title="Permalink to this headline">¶</a></h2>
<p>In most proteomics applications, a dedicated search engine (such as Comet,
Crux, Mascot, MSGFPlus, MSFragger, Myrimatch, OMSSA, SpectraST or XTandem;
all of which are supported by pyOpenMS) will be used to search data. Here, we will
use the internal <code class="docutils literal notranslate"><span class="pre">SimpleSearchEngineAlgorithm</span></code> from OpenMS used for teaching
purposes. This makes it very easy to search an (experimental) mzML file against
a fasta database of protein sequences:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">urllib.request</span> <span class="kn">import</span> <span class="n">urlretrieve</span>
<span class="c1"># from urllib import urlretrieve  # use this code for Python 2.x</span>
<span class="kn">from</span> <span class="nn">pyopenms</span> <span class="kn">import</span> <span class="o">*</span>
<span class="n">gh</span> <span class="o">=</span> <span class="s2">&quot;https://raw.githubusercontent.com/OpenMS/OpenMS/develop&quot;</span>
<span class="n">urlretrieve</span> <span class="p">(</span><span class="n">gh</span> <span class="o">+</span><span class="s2">&quot;/src/tests/topp/SimpleSearchEngine_1.mzML&quot;</span><span class="p">,</span> <span class="s2">&quot;searchfile.mzML&quot;</span><span class="p">)</span>
<span class="n">urlretrieve</span> <span class="p">(</span><span class="n">gh</span> <span class="o">+</span><span class="s2">&quot;/src/tests/topp/SimpleSearchEngine_1.fasta&quot;</span><span class="p">,</span> <span class="s2">&quot;search.fasta&quot;</span><span class="p">)</span>
<span class="n">protein_ids</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">peptide_ids</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">SimpleSearchEngineAlgorithm</span><span class="p">()</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;searchfile.mzML&quot;</span><span class="p">,</span> <span class="s2">&quot;search.fasta&quot;</span><span class="p">,</span> <span class="n">protein_ids</span><span class="p">,</span> <span class="n">peptide_ids</span><span class="p">)</span>
</pre></div>
</div>
<p>This will print search engine output including the number of peptides and
proteins in the database and how many spectra were matched to peptides and
proteins:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">Peptide</span> <span class="n">statistics</span>

  <span class="n">unmatched</span>                <span class="p">:</span> <span class="mi">0</span> <span class="p">(</span><span class="mi">0</span> <span class="o">%</span><span class="p">)</span>
  <span class="n">target</span><span class="o">/</span><span class="n">decoy</span><span class="p">:</span>
    <span class="n">match</span> <span class="n">to</span> <span class="n">target</span> <span class="n">DB</span> <span class="n">only</span><span class="p">:</span> <span class="mi">2</span> <span class="p">(</span><span class="mi">100</span> <span class="o">%</span><span class="p">)</span>
    <span class="n">match</span> <span class="n">to</span> <span class="n">decoy</span> <span class="n">DB</span> <span class="n">only</span> <span class="p">:</span> <span class="mi">0</span> <span class="p">(</span><span class="mi">0</span> <span class="o">%</span><span class="p">)</span>
    <span class="n">match</span> <span class="n">to</span> <span class="n">both</span>          <span class="p">:</span> <span class="mi">0</span> <span class="p">(</span><span class="mi">0</span> <span class="o">%</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="section" id="psm-inspection">
<h2>PSM inspection<a class="headerlink" href="#psm-inspection" title="Permalink to this headline">¶</a></h2>
<p>We can now investigate the individual hits as we have done before in the
<a class="reference external" href="datastructures_id.html#peptideidentification">Identification tutorial</a>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">peptide_id</span> <span class="ow">in</span> <span class="n">peptide_ids</span><span class="p">:</span>
  <span class="c1"># Peptide identification values</span>
  <span class="k">print</span> <span class="p">(</span><span class="mi">35</span><span class="o">*</span><span class="s2">&quot;=&quot;</span><span class="p">)</span>
  <span class="k">print</span> <span class="p">(</span><span class="s2">&quot;Peptide ID m/z:&quot;</span><span class="p">,</span> <span class="n">peptide_id</span><span class="o">.</span><span class="n">getMZ</span><span class="p">())</span>
  <span class="k">print</span> <span class="p">(</span><span class="s2">&quot;Peptide ID rt:&quot;</span><span class="p">,</span> <span class="n">peptide_id</span><span class="o">.</span><span class="n">getRT</span><span class="p">())</span>
  <span class="k">print</span> <span class="p">(</span><span class="s2">&quot;Peptide scan index:&quot;</span><span class="p">,</span> <span class="n">peptide_id</span><span class="o">.</span><span class="n">getMetaValue</span><span class="p">(</span><span class="s2">&quot;scan_index&quot;</span><span class="p">))</span>
  <span class="k">print</span> <span class="p">(</span><span class="s2">&quot;Peptide scan name:&quot;</span><span class="p">,</span> <span class="n">peptide_id</span><span class="o">.</span><span class="n">getMetaValue</span><span class="p">(</span><span class="s2">&quot;scan_index&quot;</span><span class="p">))</span>
  <span class="k">print</span> <span class="p">(</span><span class="s2">&quot;Peptide ID score type:&quot;</span><span class="p">,</span> <span class="n">peptide_id</span><span class="o">.</span><span class="n">getScoreType</span><span class="p">())</span>
  <span class="c1"># PeptideHits</span>
  <span class="k">for</span> <span class="n">hit</span> <span class="ow">in</span> <span class="n">peptide_id</span><span class="o">.</span><span class="n">getHits</span><span class="p">():</span>
    <span class="k">print</span><span class="p">(</span><span class="s2">&quot; - Peptide hit rank:&quot;</span><span class="p">,</span> <span class="n">hit</span><span class="o">.</span><span class="n">getRank</span><span class="p">())</span>
    <span class="k">print</span><span class="p">(</span><span class="s2">&quot; - Peptide hit charge:&quot;</span><span class="p">,</span> <span class="n">hit</span><span class="o">.</span><span class="n">getCharge</span><span class="p">())</span>
    <span class="k">print</span><span class="p">(</span><span class="s2">&quot; - Peptide hit sequence:&quot;</span><span class="p">,</span> <span class="n">hit</span><span class="o">.</span><span class="n">getSequence</span><span class="p">())</span>
    <span class="n">mz</span> <span class="o">=</span> <span class="n">hit</span><span class="o">.</span><span class="n">getSequence</span><span class="p">()</span><span class="o">.</span><span class="n">getMonoWeight</span><span class="p">(</span><span class="n">Residue</span><span class="o">.</span><span class="n">ResidueType</span><span class="o">.</span><span class="n">Full</span><span class="p">,</span> <span class="n">hit</span><span class="o">.</span><span class="n">getCharge</span><span class="p">())</span> <span class="o">/</span> <span class="n">hit</span><span class="o">.</span><span class="n">getCharge</span><span class="p">()</span>
    <span class="k">print</span><span class="p">(</span><span class="s2">&quot; - Peptide hit monoisotopic m/z:&quot;</span><span class="p">,</span> <span class="n">mz</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="s2">&quot; - Peptide ppm error:&quot;</span><span class="p">,</span> <span class="nb">abs</span><span class="p">(</span><span class="n">mz</span> <span class="o">-</span> <span class="n">peptide_id</span><span class="o">.</span><span class="n">getMZ</span><span class="p">())</span><span class="o">/</span><span class="n">mz</span> <span class="o">*</span><span class="mi">10</span><span class="o">**</span><span class="mi">6</span> <span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="s2">&quot; - Peptide hit score:&quot;</span><span class="p">,</span> <span class="n">hit</span><span class="o">.</span><span class="n">getScore</span><span class="p">())</span>
</pre></div>
</div>
<p>We notice that the second peptide spectrum match (PSM) was found for the third
spectrum in the file for a precursor at 775.38 m/z for the sequence
<code class="docutils literal notranslate"><span class="pre">RPGADSDIGGFGGLFDLAQAGFR</span></code>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">tsg</span> <span class="o">=</span> <span class="n">TheoreticalSpectrumGenerator</span><span class="p">()</span>
<span class="n">thspec</span> <span class="o">=</span> <span class="n">MSSpectrum</span><span class="p">()</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">Param</span><span class="p">()</span>
<span class="n">p</span><span class="o">.</span><span class="n">setValue</span><span class="p">(</span><span class="s2">&quot;add_metainfo&quot;</span><span class="p">,</span> <span class="s2">&quot;true&quot;</span><span class="p">)</span>
<span class="n">tsg</span><span class="o">.</span><span class="n">setParameters</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
<span class="n">peptide</span> <span class="o">=</span> <span class="n">AASequence</span><span class="o">.</span><span class="n">fromString</span><span class="p">(</span><span class="s2">&quot;RPGADSDIGGFGGLFDLAQAGFR&quot;</span><span class="p">)</span>
<span class="n">tsg</span><span class="o">.</span><span class="n">getSpectrum</span><span class="p">(</span><span class="n">thspec</span><span class="p">,</span> <span class="n">peptide</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
<span class="c1"># Iterate over annotated ions and their masses</span>
<span class="k">for</span> <span class="n">ion</span><span class="p">,</span> <span class="n">peak</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">thspec</span><span class="o">.</span><span class="n">getStringDataArrays</span><span class="p">()[</span><span class="mi">0</span><span class="p">],</span> <span class="n">thspec</span><span class="p">):</span>
    <span class="k">print</span><span class="p">(</span><span class="n">ion</span><span class="p">,</span> <span class="n">peak</span><span class="o">.</span><span class="n">getMZ</span><span class="p">())</span>

<span class="n">e</span> <span class="o">=</span> <span class="n">MSExperiment</span><span class="p">()</span>
<span class="n">MzMLFile</span><span class="p">()</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;searchfile.mzML&quot;</span><span class="p">,</span> <span class="n">e</span><span class="p">)</span>
<span class="k">print</span> <span class="p">(</span><span class="s2">&quot;Spectrum native id&quot;</span><span class="p">,</span> <span class="n">e</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">getNativeID</span><span class="p">()</span> <span class="p">)</span>
<span class="n">mz</span><span class="p">,</span><span class="n">i</span> <span class="o">=</span> <span class="n">e</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">get_peaks</span><span class="p">()</span>
<span class="n">peaks</span> <span class="o">=</span> <span class="p">[(</span><span class="n">mz</span><span class="p">,</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">mz</span><span class="p">,</span><span class="n">i</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">mz</span><span class="p">,</span><span class="n">i</span><span class="p">)</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span> <span class="mi">1500</span> <span class="ow">and</span> <span class="n">mz</span> <span class="o">&gt;</span> <span class="mi">300</span><span class="p">]</span>
<span class="k">for</span> <span class="n">peak</span> <span class="ow">in</span> <span class="n">peaks</span><span class="p">:</span>
  <span class="k">print</span> <span class="p">(</span><span class="n">peak</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;mz&quot;</span><span class="p">,</span> <span class="n">peak</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="s2">&quot;int&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Comparing the theoretical spectrum and the experimental spectrum for
<code class="docutils literal notranslate"><span class="pre">RPGADSDIGGFGGLFDLAQAGFR</span></code> we can easily see that the most abundant ions in the
spectrum are y8 (877.452 m/z), b10 (926.432), y9 (1024.522 m/z) and b13
(1187.544 m/z).</p>
</div>
<div class="section" id="visualization">
<h2>Visualization<a class="headerlink" href="#visualization" title="Permalink to this headline">¶</a></h2>
<p>When loading the <code class="docutils literal notranslate"><span class="pre">searchfile.mzML</span></code> into the OpenMS
visualization software TOPPView, we can convince ourselves that the observed
spectrum indeed was generated by the peptide <code class="docutils literal notranslate"><span class="pre">RPGADSDIGGFGGLFDLAQAGFR</span></code> by loading
the corresponding theoretical spectrum into the viewer using “Tools”-&gt;”Generate
theoretical spectrum”:</p>
<img alt="_images/psm.png" src="_images/psm.png" />
<p>From our output above, we notice that the second peptide spectrum match (PSM)
at 775.38 m/z for sequence <code class="docutils literal notranslate"><span class="pre">RPGADSDIGGFGGLFDLAQAGFR</span></code> was found with an error
tolerance of 2.25 ppm, therefore if we set the precursor mass tolerance to 4
ppm (+/- 2ppm), we expect that we will not find the hit at 775.38 m/z any more:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">salgo</span> <span class="o">=</span> <span class="n">SimpleSearchEngineAlgorithm</span><span class="p">()</span>
<span class="n">p</span> <span class="o">=</span> <span class="n">salgo</span><span class="o">.</span><span class="n">getDefaults</span><span class="p">()</span>
<span class="k">print</span> <span class="p">(</span> <span class="n">p</span><span class="o">.</span><span class="n">items</span><span class="p">()</span> <span class="p">)</span>
<span class="n">p</span><span class="p">[</span><span class="sa">b</span><span class="s1">&#39;precursor:mass_tolerance&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mf">4.0</span>
<span class="n">salgo</span><span class="o">.</span><span class="n">setParameters</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>

<span class="n">protein_ids</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">peptide_ids</span> <span class="o">=</span> <span class="p">[]</span>
<span class="n">salgo</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;searchfile.mzML&quot;</span><span class="p">,</span> <span class="s2">&quot;search.fasta&quot;</span><span class="p">,</span> <span class="n">protein_ids</span><span class="p">,</span> <span class="n">peptide_ids</span><span class="p">)</span>
<span class="k">print</span><span class="p">(</span><span class="s2">&quot;Found&quot;</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">peptide_ids</span><span class="p">),</span> <span class="s2">&quot;peptides&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>As we can see, using a smaller precursor mass tolerance leads the algorithm to
find only one hit instead of two. Similarly, if we use the wrong enzyme for
the digestion (e.g. <code class="docutils literal notranslate"><span class="pre">p[b'enzyme']</span> <span class="pre">=</span> <span class="pre">&quot;Formic_acid&quot;</span></code>), we find no results.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="chromatographic_analysis.html" class="btn btn-neutral float-right" title="Chromagraphic Analysis" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="feature_detection.html" class="btn btn-neutral float-left" title="Feature Detection" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2019, OpenMS Team

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>